value-processor/src/org/immutables/value/processor/Gsons.generator

```diff
@@ -133,6 +133,9 @@ public final class [typeAdaptersName] implements TypeAdapterFactory {
 [checkAttributes type]
 [for allAttributes = type.allMarshalingAttributes, t = type.typeAbstract, im = type.typeImmutable]
 
+[if type allowsClasspathAnnotation 'org.immutables.value.Generated']
+@org.immutables.value.Generated(from = "[type.typeAbstract.relativeRaw]", generator = "Gsons")
+[/if]
```

Comment:
Elsewhere the following construct is used instead:
```
[if type allowsClasspathAnnotation 'javax.annotation.processing.Generated']
@javax.annotation.processing.Generated("org.immutables.processor.ProxyProcessor")
[else if type allowsClasspathAnnotation 'javax.annotation.Generated']
@javax.annotation.Generated("org.immutables.processor.ProxyProcessor")
[/if]
```

Grepping indicates that the different approaches are used with similar frequency:
```
$ git grep javax.annotation.Generated | wc -l
27
$ git grep javax.annotation.processing.Generated | wc -l
20
$ git grep org.immutables.value.Generated | wc -l
33 
```
@elucash is the plan to use `javax.annotation.processing.Generated` or `org.immutables.value.Generated` throughout the code base?

(IIRC `javax.annotation.processing.Generated` is pitched as the JDK 9+ replacement of `javax.annotation.Generated`.)

Reply:
@Stephan202 as far as I understand it, we use `org.immutables.value.Generated` to tell tools like Jacoco during runtime what was generated. The Java types `javax.annotation.Generated` and `javax.annotation.Generated` have `@Retention(SOURCE)`.

Reply:
@before good point about `@Retention`; I didn't consider that since I'm mostly interested in `@Generated` in the context of static analysis, such as using Error Prone.

I wonder: for tools that look for a fully qualified class name (rather than any annotation with simple name `@Generated`), would it make sense to add both annotations?

Reply:
I don't know a tool yet that expects `@Generated` annotation on nested classes other then JaCoCo. Error Prone expects that everything is generated when the outer class is annotated with `javax.annotation.Generated`. I wouldn't spend time on this until we know a tool that needs it on nested classes.

Reply:
Want to clarify:

* Source retained Generated (newer version `javax.annotation.processing.Generated` or older `javax.annotation.Generated`) is only used on a top level type, everything nested is considered generated without annotation. Newer takes precedence, obviously. They are used automatically if found on compilation classpath during annotation processing and can be disabled by whitelist override in styles or blacklisting via classpath fence (extension resource in META-INF).

```java
[if type allowsClasspathAnnotation 'javax.annotation.processing.Generated']
@javax.annotation.processing.Generated("org.immutables.processor.ProxyProcessor")
[else if type allowsClasspathAnnotation 'javax.annotation.Generated']
@javax.annotation.Generated("org.immutables.processor.ProxyProcessor")
[/if]
```

* Class retained `org.immutables.value.Generated` is covering each generated type, both top level and nested, to end up in each generated `*.class` file. It is always on the classpath (our annotation API jar has `Value.*` and `Generated` packaged together), but can be turned off by whitelist override in styles.

```java
[if type allowsClasspathAnnotation 'org.immutables.value.Generated']
@org.immutables.value.Generated(from = "[type.typeAbstract.relativeRaw]", generator = "Gsons")
[/if]
```

* If someone needs some runtime retained annotation, the option is to use (at least to try) annotation injection. (see `annotate` module)